### PR TITLE
[backend/frontend] Implement full text search for documents (#1483)

### PR DESCRIPTION
**Pull request by SouadHadjiat**
### Proposed changes

PR with first version of imported files indexing and full text search for enterprise edition

* Creation of files index (opencti_files) and configuration of attachment processor
* Adding FileIndexManager module to index all imported files (global and in entities)
* In global search, adding button to extend the search to indexed files
* handle entities markings / organization restriction changes : update indexed files to have the same data restrictions

First part of file index management panel :
* Adding "File indexing" menu in settings top bar
* Display enterprise edition info if not enabled
* Display requirements for file indexing (elasticsearch / opensearch)
* Display number and size of files that will be indexed
* Manage file indexing : start / pause, display number and size of files indexed

Indexation is paused by default.

About file indexing :
We support these mime types : text/plain, text/csv, application/pdf, application/vnd.ms-excel, application/vnd.openxmlformats-officedocument.spreadsheetml.sheet (excel sheets)

### Related issues

* https://github.com/OpenCTI-Platform/opencti/issues/1483

### Checklist

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality
- [x] I wrote test cases for the relevant uses case
- [ ] I added/update the relevant documentation (either on github or on notion)
- [ ] Where necessary I refactored code to improve the overall quality

### Further comments

https://www.notion.so/filigran/Management-panel-for-File-indexing-de5a1879c62f41af99943c58a30f4aa6

https://www.notion.so/filigran/Full-text-search-for-documents-11a67658fc37410cbbbdde2db78da73a#3281f177046749678fdbb8539b5cbaab

### CODE REVIEW

File: opencti-platform/opencti-graphql/src/manager/fileIndexManager.ts
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/manager/fileIndexManager.ts]: There are things left

File: opencti-platform/opencti-graphql/src/manager/fileIndexManager.ts
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/manager/fileIndexManager.ts]: This is not possible to have this filter direclty in the S3client ? When we send the request ?

File: opencti-platform/opencti-graphql/src/manager/fileIndexManager.ts
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/manager/fileIndexManager.ts]: I don't understand why we check on this (makring, authorities)

File: opencti-platform/opencti-graphql/src/manager/fileIndexManager.ts
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/manager/fileIndexManager.ts]: Is this part not enough to index the new files? Rather than also adding stream handler management?
All files added in an entity update end up in minio.

File: opencti-platform/opencti-front/src/private/components/search/SearchIndexedFileLine.tsx
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/search/SearchIndexedFileLine.tsx:57]: Better to use ListItemButton

File: opencti-platform/opencti-front/src/private/components/search/SearchIndexedFiles.tsx
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/search/SearchIndexedFiles.tsx]: Can you use render here ?

File: opencti-platform/opencti-front/src/private/components/search/SearchIndexedFileLine.tsx
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/search/SearchIndexedFileLine.tsx:70]: You can use render in SearchIndexedFiles.tsx to simplify this code

**Comment by Archidoit** [comment]
Is it usefull to keep the 'line view' button on the right since there is no other view possibility ? 
image

**Comment by Archidoit** [comment]
If we click again on 'Extends this search to indexed files', it would be great that the panel closes. 
image

**Comment by Archidoit** [comment]
Sorting file result (by name or 'attached entity type' is not working : 
image

Note: Would be great to also be able to sort by the number of occurences
@Jipegien

**Comment by Archidoit** [comment]
What are all the files extensions for which this feature is working ?
I uploaded a txt file containing the world 'filter' and it was not found... But an other more simple txt file was found...

File: opencti-platform/opencti-graphql/src/manager/fileIndexManager.ts
Comment (SouadHadjiat) [reply opencti-platform/opencti-graphql/src/manager/fileIndexManager.ts]: no there is no filter on mime types in listObjects query

File: opencti-platform/opencti-graphql/src/manager/fileIndexManager.ts
Comment (SouadHadjiat) [reply opencti-platform/opencti-graphql/src/manager/fileIndexManager.ts]: what do you mean ? The manager has two parts : one that indexes all new uploaded files in minio, another one that listens to the stream to update indexed files linked to entities if markings / organization restrictions changes.

File: opencti-platform/opencti-front/src/private/components/search/SearchIndexedFiles.tsx
Comment (SouadHadjiat) [reply opencti-platform/opencti-front/src/private/components/search/SearchIndexedFiles.tsx]: it is not sortable for now, it should be false here.

**Comment by SouadHadjiat** [comment]
> What are all the files extensions for which this feature is working ?

@Archidoit We support these mime types : text/plain, text/csv, application/pdf, application/vnd.ms-excel, application/vnd.openxmlformats-officedocument.spreadsheetml.sheet (excel sheets). you can find them in default.json

**Comment by SouadHadjiat** [comment]
> Sorting file result (by name or 'attached entity type' is not working : image
> 
> Note: Would be great to also be able to sort by the number of occurences

It's a mistake, we don't support custom sorting for now (it will be in V2), we forgot to remove column sorting. And sorting by number of occurrence won't be possible, because it's a data we compute, not something stored that we could sort on.

File: opencti-platform/opencti-graphql/src/manager/fileIndexManager.ts
Comment (SouadHadjiat) [reply opencti-platform/opencti-graphql/src/manager/fileIndexManager.ts]: we listen to the stream and filter events that are only updates on markings / organization restrictions.
authorized members and authorities are not handled yet, the comment should be : `// test if markings or organization sharing have been updated`

File: opencti-platform/opencti-front/src/private/components/search/SearchIndexedFiles.tsx
Comment (RomuDeuxfois) [reply opencti-platform/opencti-front/src/private/components/search/SearchIndexedFiles.tsx]: I mean the render property

      isSortable: true,
      render: (node: Organization) => node.name,

File: opencti-platform/opencti-graphql/src/database/engine.js
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/database/engine.js:750]: Awesome !

File: opencti-platform/opencti-graphql/src/database/file-storage.js
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/database/file-storage.js]: Why this changes ?

**Comment by Jipegien** [comment]
> Sorting file result (by name or 'attached entity type' is not working : image
> 
> Note: Would be great to also be able to sort by the number of occurences @Jipegien

Agreed. Should be able to sort by filename, date, entity type and number of occurences.

File: opencti-platform/opencti-graphql/src/database/file-storage.js
Comment (SouadHadjiat) [reply opencti-platform/opencti-graphql/src/database/file-storage.js]: for typescript typing

**Comment by SouadHadjiat** [comment]
> > Sorting file result (by name or 'attached entity type' is not working : image
> > Note: Would be great to also be able to sort by the number of occurences @Jipegien
> 
> Agreed. Should be able to sort by filename, date, entity type and number of occurences.

@Jipegien it's not possible to sort by number of occurrence. About other sorting, we can only sort by filename, and uploadDate, not by entity_type, or entity_name. for V2

**Comment by SouadHadjiat** [comment]
**PR status** 

- waiting for more reviews and tests.
- some review comments have already been applied in latest commits. There are still a few to work on.

File: opencti-platform/opencti-graphql/src/database/file-storage.js
Comment (RomuDeuxfois) [reply opencti-platform/opencti-graphql/src/database/file-storage.js]: Ok, but we reverse the properties order.
Is that ok ?

**Review by RomuDeuxfois — APPROVED** [review]

**Comment by Kedae** [comment]
We need to test perf before merging

File: opencti-platform/opencti-graphql/src/manager/fileIndexManager.ts
Comment (Kedae) [inline opencti-platform/opencti-graphql/src/manager/fileIndexManager.ts:16]: We need to test perf before merging

**Review by Kedae — DISMISSED** [review]

**Comment by SouadHadjiat** [comment]
FileIndexManager is now disabled by default in configuration (default.json), PR is ready to be merged when master is stable @Kedae @richard-julien

**Comment by SouadHadjiat** [comment]
PR ready to be merged, `fileIndexManager` is now completely disabled @richard-julien @Kedae

**=> PR status updated to draft** : we need to change FileIndexStatus to ManagerConfiguration, add a start / pause logic, have a first version of file index management panel.

**Comment by SouadHadjiat** [comment]
**PR status**
Ready for review

**Pull request merged**
